File: rbm.py
import numpy as np


class RBM:
    def __init__(self, num_visible, num_hidden):
        self.num_hidden = num_hidden
        self.num_visible = num_visible
        self.debug_print = True

        # Initialize a weight matrix, of dimensions (num_visible x num_hidden), using
        # a uniform distribution between -sqrt(6. / (num_hidden + num_visible))
        # and sqrt(6. / (num_hidden + num_visible)). One could vary the
        # standard deviation by multiplying the interval with appropriate value.
        # Here we initialize the weights with mean 0 and standard deviation 0.1.
        # Reference: Understanding the difficulty of training deep feedforward
        # neural networks by Xavier Glorot and Yoshua Bengio
        np_rng = np.random.RandomState(1235)

        self.weights = np.asarray(
            np_rng.uniform(
                low=-0.1 * np.sqrt(6.0 / (num_hidden + num_visible)),
                high=0.1 * np.sqrt(6.0 / (num_hidden + num_visible)),
                size=(num_visible, num_hidden),
            )
        )

        # Insert weights for the bias units into the first row and first column.
        self.weights = np.insert(self.weights, 0, 0, axis=0)
        self.weights = np.insert(self.weights, 0, 0, axis=1)

    def train(self, data, max_epochs=1000, learning_rate=0.1):
        """
    Train the machine.

    Parameters
    ----------
    data: A matrix where each row is a training example consisting of the states of visible units.    
    """

        num_examples = data.shape[0]

        # Insert bias units of 1 into the first column.
        data = np.insert(data, 0, 1, axis=1)
        print(data)

        for epoch in range(max_epochs):
            # Clamp to the data and sample from the hidden units.
            # (This is the "positive CD phase", aka the reality phase.)
            print(f"Weights: {self.weights.shape}")
            pos_hidden_activations = np.dot(data, self.weights)
            print(f"activations: {pos_hidden_activations.shape}")
            pos_hidden_probs = self._logistic(pos_hidden_activations)
            pos_hidden_probs[:, 0] = 1  # Fix the bias unit.
            print(pos_hidden_probs)
            pos_hidden_states = pos_hidden_probs > np.random.rand(
                num_examples, self.num_hidden + 1
            )
            print(pos_hidden_states)
            # Note that we're using the activation *probabilities* of the hidden states, not the hidden states
            # themselves, when computing associations. We could also use the states; see section 3 of Hinton's
            # "A Practical Guide to Training Restricted Boltzmann Machines" for more.
            pos_associations = np.dot(data.T, pos_hidden_probs)
            print(pos_associations)

            # Reconstruct the visible units and sample again from the hidden units.
            # (This is the "negative CD phase", aka the daydreaming phase.)
            neg_visible_activations = np.dot(pos_hidden_states, self.weights.T)
            neg_visible_probs = self._logistic(neg_visible_activations)
            neg_visible_probs[:, 0] = 1  # Fix the bias unit.
            neg_hidden_activations = np.dot(neg_visible_probs, self.weights)
            neg_hidden_probs = self._logistic(neg_hidden_activations)
            # Note, again, that we're using the activation *probabilities* when computing associations, not the states
            # themselves.
            neg_associations = np.dot(neg_visible_probs.T, neg_hidden_probs)

            # Update weights.
            self.weights += learning_rate * (
                (pos_associations - neg_associations) / num_examples
            )

            error = np.sum((data - neg_visible_probs) ** 2)
            if self.debug_print:
                print("Epoch %s: error is %s" % (epoch, error))

    def run_visible(self, data):
        """
        Assuming the RBM has been trained (so that weights for the network have been learned),
        run the network on a set of visible units, to get a sample of the hidden units.
        
        Parameters
        ----------
        data: A matrix where each row consists of the states of the visible units.
        
        Returns
        -------
        hidden_states: A matrix where each row consists of the hidden units activated from the visible
        units in the data matrix passed in.
        """
        
        num_examples = data.shape[0]
        
        # Create a matrix, where each row is to be the hidden units (plus a bias unit)
        # sampled from a training example.
        hidden_states = np.ones((num_examples, self.num_hidden + 1))
        
        # Insert bias units of 1 into the first column of data.
        data = np.insert(data, 0, 1, axis = 1)

        # Calculate the activations of the hidden units.
        hidden_activations = np.dot(data, self.weights)
        # Calculate the probabilities of turning the hidden units on.
        hidden_probs = self._logistic(hidden_activations)
        # Turn the hidden units on with their specified probabilities.
        hidden_states[:,:] = hidden_probs > np.random.rand(num_examples, self.num_hidden + 1)
        # Always fix the bias unit to 1.
        # hidden_states[:,0] = 1
    
        # Ignore the bias units.
        hidden_states = hidden_states[:,1:]
        return hidden_states

    def _logistic(self, x):
        return 1.0 / (1 + np.exp(-x))


if __name__ == "__main__":
    r = RBM(num_visible=6, num_hidden=2)
    training_data = np.array(
        [
            [1, 1, 1, 0, 0, 0],
            [1, 0, 1, 0, 0, 0],
            [1, 1, 1, 0, 0, 0],
            [0, 0, 1, 1, 1, 0],
            [0, 0, 1, 1, 0, 0],
            [0, 0, 1, 1, 1, 0],
        ]
    )
    r.train(training_data, max_epochs=1)
    print(r.weights)
    user = np.array([[0, 0, 0, 1, 1, 0]])
    print(r.run_visible(user))

File: scripts/rbm.jl
using Random
using RandomNumbers.Xorshifts
using LinearAlgebra


rng = Xorshift1024Plus(21457)

mutable struct BoltzmannMachine
    W::AbstractArray
    rng::AbstractRNG
end

function init!(v, h; rng = nothing)
    w = Matrix{Float64}(undef, v, h)
    # @show w
    # Glorot initialization for weights
    low = -0.1 * sqrt(6.0 / (h + v))
    rand!(rng, w, [low, -low])

    # Insert weights for the bias units into the first row and first column.
    w = vcat(zeros(1, h), w)
    w = hcat(zeros(v+1), w)

    return BoltzmannMachine(w, rng)
end

sigmoid(x) = 1 / (1 + exp(-x))

to_one(x) = x == true ? 1 : 0

function kstep(W; k = 50, rng = nothing)
    (N, M) = size(W)
    samples = rand(rng, 1, N)
    neg_associations = Matrix{Float64}(undef, N, M)

    for i = 1:k-1
        hidden_activations = samples * W
        # @show size(hidden_activations)
        hidden_probs = sigmoid.(hidden_activations)
        # @show size(hidden_probs)
        hidden_states = hidden_probs .> rand(rng, (1, M))
        hidden_states = to_one.(hidden_states)
        hidden_states[:, 1] .= 1

        visible_activations = hidden_states * W'
        visible_probs = sigmoid.(visible_activations)
        # @show size(visible_probs)
        # @show size(visible_probs')
        # @show size(hidden_probs)
        neg_associations = visible_probs' * hidden_probs
        # @show size(neg_associations)
    end

    return neg_associations
end

function train(data, rbm; epochs = 3000, η = 0.001)
    (N, M) = size(data)

    # Add bias to data
    data_bias = hcat(ones(N), data)

    neg_associations = Matrix{Float64}(undef, N, M)

    for e = 1:epochs
        pos_hidden_activations = data_bias * rbm.W
        pos_hidden_probs = sigmoid.(pos_hidden_activations)
        # Fix the bias unit
        pos_hidden_probs[:, 1] .= 1
        pos_hidden_states = pos_hidden_probs .> rand(rbm.rng, size(pos_hidden_probs)...)
        pos_hidden_states = to_one.(pos_hidden_states)
        pos_associations = data_bias' * pos_hidden_probs
        # @show size(pos_associations)

        neg_associations = kstep(rbm.W; k = 5000, rng = rbm.rng)

        # Update weights.
        @. rbm.W += η * ((pos_associations - neg_associations) / N)
    end
end

function sample(rbm; k = 50, rng = nothing)
    (N, M) = size(rbm.W)
    samples = Array{Float64}(undef, k, N, N)
    # samples = Matrix{Float64}(undef, k, N)
    samples[1, :, :] = rand(rng, 1, N, N)

    for i = 1:k-1
        hidden_activations = samples[i, :, :] * rbm.W
        hidden_probs = sigmoid.(hidden_activations)
        hidden_probs[:, 1] .= 1
        hidden_states = hidden_probs .> rand(rng, 1, M)
        hidden_states = to_one.(hidden_states)

        visible_activations = hidden_states * rbm.W'
        # @show size(visible_activations)
        visible_probs = sigmoid.(visible_activations)
        samples[i+1, :, :] = visible_probs
    end

    return samples
end

data = [
    [1, 1, 1, 0, 0, 0];
    [1, 0, 1, 0, 0, 0];
    [1, 1, 1, 0, 0, 0];
    [0, 0, 1, 1, 1, 0];
    [0, 0, 1, 1, 0, 0];
    [0, 0, 1, 1, 1, 0]
]
data = reshape(data, (6, 6))

rbm = init!(6, 2; rng = rng)

train(data, rbm)

new_data = sample(rbm; k = 2000, rng = rng)
display(new_data[end, :, :])
